share/man/man4/filemon.4
.\"	$NetBSD$
.\"
.\" Copyright (c) 2011, Juniper Networks, Inc.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
.\" "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
.\" LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
.\" A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
.\" OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
.\" SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
.\" LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
.\" OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd January 6, 2016
.Dt FILEMON 4
.Os
.Sh NAME
.Nm filemon
.Nd track interesting system calls
.Sh SYNOPSIS
.Cd "pseudo-device filemon"
.Sh DESCRIPTION
In normal situations,
.Nm
is not built-in to the kernel, and a call to open
.Pa /dev/filemon
will auto-load the
.Nm
module (see
.Xr module 7
for more details).
.Pp
(Although not recommended, the
.Nm
facility can be included in a kernel build by adding
.Bd -literal -offset indent
pseudo-device filemon
.Ed
.Pp
to the kernel configuration file.)
.Pp
.Nm
provides a means for tracking the successful system calls performed by
a process and its descendants.
It is used by
.Xr make 1
to track the activities of build scripts, for the purpose of automatically
learning dependencies.
.Pp
The data captured by
.Nm
for the script
.Bd -literal -offset indent
n=`wc -l /etc/motd`; echo "int motd_lines = $n;" > foo.h.new
cmp -s foo.h foo.h.new 2> /dev/null || mv foo.h.new foo.h
.Ed
.Pp
looks like:
.Bd -literal -offset indent
# filemon version 4
# Target pid 24291
V 4
E 29676 /bin/sh
R 29676 /etc/ld.so.conf
R 29676 /lib/libedit.so.2
R 29676 /lib/libterminfo.so.1
R 29676 /lib/libc.so.12
F 29676 4899
E 4899 /usr/bin/wc
R 4899 /etc/ld.so.conf
R 4899 /usr/lib/libc.so.12
R 4899 /etc/motd
X 4899 0
W 29676 foo.h.new
X 29676 0
# Bye bye
E 3250 /bin/sh
R 3250 /etc/ld.so.conf
R 3250 /lib/libedit.so.2
R 3250 /lib/libterminfo.so.1
R 3250 /lib/libc.so.12
W 26673 /dev/null
E 26673 /usr/bin/cmp
R 26673 /etc/ld.so.conf
R 26673 /usr/lib/libc.so.12
X 26673 2
E 576 /bin/mv
R 576 /etc/ld.so.conf
R 576 /lib/libc.so.12
M 576 'foo.h.new' 'foo.h'
X 576 0
X 3250 0
# Bye bye
.Ed
.Pp
Most records follow the format:
.Bd -literal -offset indent
type pid data
.Ed
.Pp
where
.Ar type
is one of the list below, and unless otherwise specified,
.Ar data
is a pathname.
.Bl -tag -width Ds -offset indent
.It Dv C
.Xr chdir 2 .
.It Dv D
.Xr unlink 2 .
.It Dv E
.Xr exec 3 .
.It Dv F
.Xr fork 2 ,
.Xr vfork 2 ;
.Ar data
is the process id of the child.
.It Dv L
.Xr link 2 ,
.Xr symlink 2 ;
.Ar data
is two pathnames.
.It Dv M
.Xr rename 2 ;
.Ar data
is two pathnames.
.It Dv R
.Xr open 2
for read or read-write.
.It Dv W
.Xr open 2
for writing or read-write.
.It Dv X
.Xr exit 3 ;
.Ar data
is the exit status.
.It Dv V
indicates the version of
.Nm .
.El
.Pp
A
.Nm
instance is created by opening
.Pa /dev/filemon .
Then use
.Fn ioctl filemon_fd FILEMON_SET_PID &pid
to identify the target process to monitor, and
.Fn ioctl filemon_fd FILEMON_SET_FD &output_fd
to direct the event log to an already-opened output file.
.Sh FILES
.Bd -literal
/dev/filemon
.Ed
.Sh EXAMPLES
The following example demonstrates the basic usage of
.Nm :
.Pp
.Bd -literal -offset indent
#include <filemon.h>

pid_t pid;
int filemon_fd, temp_fd;
int status;

filemon_fd = open("/dev/filemon", O_RDWR);
temp_fd = mkstemp("/tmp/filemon.XXXXXXX");
/* give filemon the temp file to use */
ioctl(filemon_fd, FILEMON_SET_FD, &temp_fd);
/* children do not need these once they exec */
fcntl(filemon_fd, F_SETFD, FD_CLOEXEC);
fcntl(temp_fd, F_SETFD, FD_CLOEXEC);

pid = fork();
switch(pid) {
 case -1:
     err(1, "cannot fork");
     break;
 case 0:
     pid = getpid();
     /* tell filemon to monitor this process */
     ioctl(filemon_fd, FILEMON_SET_PID, &pid);
     execvp(...);
     _exit(1);
     break;
 default:
     status = wait();
     close(filemon_fd);
     lseek(temp_fd, SEEK_SET, 0);
     /* read the captured syscalls from temp_fd */
     close(temp_fd);
     break;
}
.Ed
.Pp
The output of
.Nm
is intended to be simple to parse.
It is possible to achieve almost equivalent results with
.Xr dtrace 1
though on many systems this requires elevated privileges.
Also,
.Xr ktrace 1
can capture similar data, but records failed system calls as well as
successful, and is thus more complex to post-process.
.Sh HISTORY
.Nm
was contributed by Juniper Networks.
.Sh SECURITY CONSIDERATIONS
If the monitored process exits, and its pid gets reused,
.Nm
will continue to report events for the new process (and its
descendants) without any authorization checks.
.Pp
Monitoring of a process enables the target process to write to the
tracking process's file descriptor.
.Sh RESTRICTIONS
The
.Nm
facility can only be used to track processes running in the system's
native emulation.
Neither processes using any of the
.Dv COMPAT_xxx
compatibility layers nor
any descendants of such processes can be tracked.
.Pp
If two processes are monitored, and one is a descendant of the other, events
related to the descendant process and its further descendants are delivered
only to the descendant process's monitor.
If a process is being monitored by two instances of
.Nm ,
events will be delivered only to the first instance created (when
.Pa /dev/filemon
was opened), regardless of the order in which the monitoring processes
called
.Fn ioctl fd FILEMON_SET_PID pid .


doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision$>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------


Changes from NetBSD 7.0 to NetBSD 8.0:
	libc: Update to tzcode2014f. [christos 20140815]
	gmake:	Update to gmake-3.81 in external/gpl2/gmake and remove 
		gnu/dist/gmake (3.80) [christos 20140818]
	module(7): Create MODULAR_DEFAULT_AUTOLOAD config flag to control
		whether module autoloading is on by default.
		[jnemeth 20140824]
	kernel: Option SYMTAB_SPACE has been replaced with COPY_SYMTAB.
		The related changes to src/sys/conf require a rebuild of
		config(1). [joerg 20140824]
	zoneinfo: Import tzdata2014g. [apb 20140831]
	wm(4): Add internal SERDES support for 82575 and newer controllers.
		[msaitoh 20140903]
	or1k: Add new port.  [matt 20140903]
	awindma(4): Add support for Allwinner A10/A20 DMA controller.
		[jmcneill 20140906]
	awinac(4): Add support for Allwinner A10/A20 audio codec.
		[jmcneill 20140906]
	awinrtc(4): Add support for Allwinner A10/A20 real-time clock.
		[jmcneill 20140907]
	awinmmc(4): Use DMA for MMC transfers. [jmcneill 20140908]
	axp20x(4): Add AXP20x PMU driver. [jmcneill 20140909]
	rtsol(8): Removed in favour of dhcpcd. [roy 20140911]
	rtsold(8): Removed in favour of dhcpcd. [roy 20140911]
	dhcpcd(8): Import dhcpcd-6.4.5. [roy 20140918]
	arm: Add support for i.MX6 SoC. [ryo 20140925]
	gpt(8): Completed overhaul, including adding follwing subcommands:
		resize, set, unset, backup, restore, and resizedisk.
		[jnemeth 20140926]
	dhcpcd(8): Import dhcpcd-6.4.7. [roy 20140927]
	mpt(4): added bio(4) support. [jmcneill 20140927]
	gpt(8): added type subcommand to modify partition type.
		[jnemeth 20140927]
	zoneinfo: Import tzdata2014h. [apb 20141004]
	dhcpcd(8): Import dhcpcd-6.5.0. [roy 20141006]
	libc: Update to tzcode2014h. [christos 20141007]
	file(1): Upgraded to 5.20. [christos 20141010]
	arm: Add support for Allwinner A31 SoC. [jmcneill 20141010]
	awiniic(4): Add support for Allwinner A31 SoC. [jmcneill 20141012]
	awinp2wi(4): Add support for Allwinner A31 Push-Pull Two Wire
		Interface. [jmcneill 20141012]
	axp22x(4): Add AXP22x PMU driver. [jmcneill 20141012]
	awinrtc(4): Add support for Allwinner A31 real-time clock.
		[jmcneill 20141012]
	awindma(4): Add support for Allwinner A31 DMA controller.
		[jmcneill 20141013]
	awinac(4): Add support for Allwinner A31 audio codec.
		[jmcneill 20141013]
	motg(4): Add support for Allwinner A10/A20 SoC. [jmcneill 20141015]
	wpa: Import wpa_supplicant and hostapd 2.3. [christos 20141016]
	dhcpcd(8): Import dhcpcd-6.5.1. [roy 20141018]
	OpenSSH: Imported 6.7. [christos 20141018]
	resolvconf(8): Import openresolv-3.6.0 [roy 20141020]
	gcc: Support -fsanitize=address [christos 20141022]
	zoneinfo: Import tzdata2014i. [apb 20141023]
	libc: Update to tzcode2014i. [christos 20141023]
	openpam(3): update to 20140912 (ourouparia) [christos 20141024]
	pppd(8): updated to version 2.4.7. [christos 20141025]
	acpi(4): Updated ACPICA to 20140926. [christos 20141025]
	dhcpcd(8): Import dhcpcd-6.6.0 [roy 20141029]
	resolvconf(8): Import openresolv-3.6.1 [roy 20141029]
	rtsx(4): Add support for Realtek RTS5227/RTL8402/RTL8411/RTL8411B.
		[nonaka 20141029]
	flex(1): Import flex-2.5.37 [christos 20141029]
	iwn(4): Add support for Intel Centrino Wireless-N
		100/105/130/135/2200/2230. From OpenBSD. [nonaka 20141030]
	awge(4): Add support for gigabit ethernet found on various
		Allwinner SOCs. [martin 20141101]
	dhcpcd(8): Import dhcpcd-6.6.1. [roy 20141107]
	awinhdmi(4): Add support for Allwinner A20/A31 HDMI controller.
		[jmcneill 20141110]
	awintcon(4): Add support for Allwinner A20/A31 TV/LCD controller.
		[jmcneill 20141110]
	awindebe(4): Add support for Allwinner A20/A31 Display engine
		backend (DE-BE). [jmcneill 20141110]
	awinhdmiaudio(4): Add support for Allwinner A20/A31 HDMI audio
		controller. [jmcneill 20141111]
	dhcpcd(8): Import dhcpcd-6.6.2. [roy 20141114]
	awinir(4): Add support for Allwinner A20/A31 IR receiver.
		[jmcneill 20141115]
	zoneinfo: Import tzdata2014j. [apb 20141116]
	libc: Change arc4random(3) to use ChaCha20 and per-thread state.
		[riastradh 20141116]
	libc: Update to tzcode2014j. [christos 20141117]
	libpcap: Import 1.6.2. [christos 20141119]
	tcpdump(8): Import 4.6.2. [christos 20141119]
	dhcpcd(8): Import dhcpcd-6.6.4. [roy 20141126]
	dhcpcd(8): Import dhcpcd-6.6.5. [roy 20141209]
	bind: Import version 9.10.1-P1. [christos 20141209]
	dhcpcd(8): Import dhcpcd-6.6.6. [roy 20141217]
	powerpc: Switch to GCC 4.8. [matt 20141218]
	dhcpcd(8): Import dhcpcd-6.6.7. [roy 20141219]
	ntp: Import ntp 4.2.8. [christos 20141219]
	midirecord(1): Add program to record MIDI files.  [mrg 20141229]
	file: update to 5.22 security limits [christos 20150102]
	byacc: update to 20141128 for more bison compat [christos 20150103]
	mgx: driver for Southland Media Systems (now Quantum 3D) MGX SBus
		graphics devices [macallan 20150106]
	gcc: Import GCC 4.8.4.  [mrg 20150109]
	openssl: Import openssl 1.0.1k [spz 20150113]
	am-utils: Upgrade to 6.2; adds nfsv4, lustre (which we can't use)
		nfs3 rpc support for amfs [christos 20150117]

sys/dev/filemon/filemon.c
/*      $NetBSD$ */
/*
 * Copyright (c) 2010, Juniper Networks, Inc.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <sys/cdefs.h>
__KERNEL_RCSID(0, "$NetBSD$");

#include <sys/param.h>
#include <sys/kernel.h>
#include <sys/module.h>
#include <sys/conf.h>
#include <sys/file.h>
#include <sys/filedesc.h>
#include <sys/systm.h>
#include <sys/buf.h>
#include <sys/fcntl.h>
#include <sys/rwlock.h>
#include <sys/condvar.h>
#include <sys/lwp.h>
#include <sys/proc.h>
#include <sys/kmem.h>
#include <sys/syslog.h>
#include <sys/kauth.h>

#include "filemon.h"
#include "ioconf.h"

MODULE(MODULE_CLASS_DRIVER, filemon, NULL);

static dev_type_open(filemon_open);

struct cdevsw filemon_cdevsw = {
	.d_open = filemon_open,
	.d_close = noclose,
	.d_read = noread,
	.d_write = nowrite,
	.d_ioctl = noioctl,
	.d_stop = nostop,
	.d_tty = notty,
	.d_poll = nopoll,
	.d_mmap = nommap,
	.d_kqfilter = nokqfilter,
	.d_discard = nodiscard,
	.d_flag = D_MPSAFE
};

static int filemon_ioctl(struct file *, u_long, void *);
static int filemon_close(struct file *);

static const struct fileops filemon_fileops = {
	.fo_ioctl = filemon_ioctl,
	.fo_close = filemon_close,
	.fo_read = fbadop_read,
	.fo_write = fbadop_write,
	.fo_fcntl = fnullop_fcntl,
	.fo_poll = fnullop_poll,
	.fo_stat = fbadop_stat,
	.fo_kqfilter = fnullop_kqfilter,
};

static krwlock_t filemon_mtx;

static TAILQ_HEAD(, filemon) filemons_inuse =
	TAILQ_HEAD_INITIALIZER(filemons_inuse);

#ifdef DEBUG
static int logLevel = LOG_DEBUG;
#endif

void
filemon_output(struct filemon * filemon, char *msg, size_t len)
{
	struct uio auio;
	struct iovec aiov;

	if (filemon->fm_fp == NULL)
		return;

	aiov.iov_base = msg;
	aiov.iov_len = len;
	auio.uio_iov = &aiov;
	auio.uio_iovcnt = 1;
	auio.uio_resid = len;
	auio.uio_rw = UIO_WRITE;
	auio.uio_offset = (off_t) - 1;
	uio_setup_sysspace(&auio);

#ifdef DEBUG
	{
		char *cp;
		int x = 16;

		cp = strchr(msg, '\n');
		if (cp && cp - msg <= 16)
			x = (cp - msg) - 2;
		log(logLevel, "filemon_output:('%.*s%s'", x,
		    (x < 16) ? "..." : "", msg);
	}
#endif
	(*filemon->fm_fp->f_ops->fo_write) (filemon->fm_fp,
	    &(filemon->fm_fp->f_offset),
	    &auio, curlwp->l_cred, FOF_UPDATE_OFFSET);
}

void
filemon_printf(struct filemon *filemon, const char *fmt, ...)
{
	size_t len;
	va_list ap;

	va_start(ap, fmt);
	len = vsnprintf(filemon->fm_msgbufr, sizeof(filemon->fm_msgbufr),
	    fmt, ap);
	va_end(ap);
	if (len > sizeof(filemon->fm_msgbufr))
		len = sizeof(filemon->fm_msgbufr);
	filemon_output(filemon, filemon->fm_msgbufr, len);
}

static void
filemon_comment(struct filemon * filemon)
{

	filemon_printf(filemon, "# filemon version %d\n# Target pid %d\nV %d\n",
	   FILEMON_VERSION, curproc->p_pid, FILEMON_VERSION);
}


static struct filemon *
filemon_pid_check(struct proc * p)
{
	struct filemon *filemon;
	struct proc * lp;

	KASSERT(p != NULL);
	if (!TAILQ_EMPTY(&filemons_inuse)) {
		/*
		 * make sure p cannot exit
		 * until we have moved on to p_pptr
		 */
		rw_enter(&p->p_reflock, RW_READER);
		while (p) {
			TAILQ_FOREACH(filemon, &filemons_inuse, fm_link) {
				if (p->p_pid == filemon->fm_pid) {
					rw_exit(&p->p_reflock);
					return (filemon);
				}
			}
			lp = p;
			p = p->p_pptr;

			/* lock parent before releasing child */
			if (p != NULL)
				rw_enter(&p->p_reflock, RW_READER);
			rw_exit(&lp->p_reflock);
		}
	}
	return (NULL);
}

/*
 * return exclusive access to a filemon struct
 */
struct filemon *
filemon_lookup(struct proc * p)
{
	struct filemon *filemon;

	rw_enter(&filemon_mtx, RW_READER);
	filemon = filemon_pid_check(p);
	if (filemon) {
		rw_enter(&filemon->fm_mtx, RW_WRITER);
	}
	rw_exit(&filemon_mtx);
	return filemon;
}

static struct filemon *
filemon_fp_data(struct file * fp, int lck)
{
	struct filemon *filemon;
	
	rw_enter(&filemon_mtx, RW_READER);
	filemon = fp->f_data;
	if (filemon && lck) {
		rw_enter(&filemon->fm_mtx, lck);
	}
	rw_exit(&filemon_mtx);
	return filemon;
}

static int n_open = 0;

static int
filemon_open(dev_t dev, int oflags __unused, int mode __unused,
    struct lwp * l __unused)
{
	struct filemon *filemon;
	struct file *fp;
	int error, fd;

	/* falloc() will fill in the descriptor for us. */
	if ((error = fd_allocfile(&fp, &fd)) != 0)
		return error;

	filemon = kmem_alloc(sizeof(struct filemon), KM_SLEEP);
	rw_init(&filemon->fm_mtx);
	filemon->fm_fp = NULL;
	filemon->fm_pid = curproc->p_pid;

	rw_enter(&filemon_mtx, RW_WRITER);
	TAILQ_INSERT_TAIL(&filemons_inuse, filemon, fm_link);
	n_open++;
	rw_exit(&filemon_mtx);

	return fd_clone(fp, fd, oflags, &filemon_fileops, filemon);
}


static int
filemon_close(struct file * fp)
{
	struct filemon *filemon;

#ifdef DEBUG
	log(logLevel, "filemon_close()");
#endif
	/*
	 * Follow the same lock order as filemon_lookup()
	 * and filemon_fp_data() but hold exclusive access to
	 * filemon_mtx until we are done.
	 */
	rw_enter(&filemon_mtx, RW_WRITER);
	filemon = fp->f_data;
	if (!filemon) {
		rw_exit(&filemon_mtx);
		return EBADF;
	}
	/* ensure that filemon_lookup() will now fail */
	TAILQ_REMOVE(&filemons_inuse, filemon, fm_link);
	n_open--;
	/* ensure that filemon_fp_data() will now fail */
	fp->f_data = NULL;

	/*
	 * once we have exclusive access, it should never be used again
	 */
	rw_enter(&filemon->fm_mtx, RW_WRITER);
	if (filemon->fm_fp) {
		closef(filemon->fm_fp);	/* release our reference */
		filemon->fm_fp = NULL;
	}
	rw_exit(&filemon->fm_mtx);
	rw_destroy(&filemon->fm_mtx);
	kmem_free(filemon, sizeof(struct filemon));
	rw_exit(&filemon_mtx);
	return (0);
}

static int
filemon_ioctl(struct file * fp, u_long cmd, void *data)
{
	int error = 0;
	int fd;
	struct filemon *filemon;
	struct proc *tp;

#ifdef DEBUG
	log(logLevel, "filemon_ioctl(%lu)", cmd);;
#endif

	/*
	 * this ensures we cannot get filemon if it is closing.
	 */
	filemon = filemon_fp_data(fp, RW_WRITER);
	if (!filemon)
		return EBADF;

	/* filemon_fp_data() has locked the entry - make sure to unlock! */

	switch (cmd) {
	case FILEMON_SET_FD:
		/* Set the output file descriptor. */

		/* First, release any current output file descriptor */
		if (filemon->fm_fp)
			closef(filemon->fm_fp);

		/* Now set up the new one */
		fd = *((int *) data);
		if ((filemon->fm_fp = fd_getfile2(curproc, fd)) == NULL) {
			error = EBADF;
			break;
		}
		/* Write the file header. */
		filemon_comment(filemon);
		break;

	case FILEMON_SET_PID:
		/* Set the monitored process ID - if allowed. */
		mutex_enter(proc_lock);
		tp = proc_find(*((pid_t *) data));
		if (tp == NULL ||
		    tp->p_emul != &emul_netbsd) {
			error = ESRCH;
			mutex_exit(proc_lock);
			break;
		}

		error = kauth_authorize_process(curproc->p_cred,
		    KAUTH_PROCESS_CANSEE, tp,
		    KAUTH_ARG(KAUTH_REQ_PROCESS_CANSEE_ENTRY), NULL, NULL);
		if (!error) {
			filemon->fm_pid = tp->p_pid;
		}
		mutex_exit(proc_lock);
		break;

	default:
		error = EINVAL;
		break;
	}

	rw_exit(&filemon->fm_mtx);
	return (error);
}

static int
filemon_load(void *dummy __unused)
{
	rw_init(&filemon_mtx);

	/* Install the syscall wrappers. */
	return filemon_wrapper_install();
}

/*
 * If this gets called we are linked into the kernel
 */
void
filemonattach(int num)
{

	/*
	 * Don't call filemon_load() here - it will be called from
	 * filemon_modcmd() during module initialization.
	 */
#if 0
	filemon_load(NULL);
#endif
}


static int
filemon_unload(void)
{
	int error = 0;

	rw_enter(&filemon_mtx, RW_WRITER);

	if (TAILQ_FIRST(&filemons_inuse) != NULL)
		error = EBUSY;
	else {
		/* Deinstall the syscall wrappers. */
		error = filemon_wrapper_deinstall();
	}
	rw_exit(&filemon_mtx);

	if (error == 0) {
		rw_destroy(&filemon_mtx);
	}
	return (error);
}

static int
filemon_modcmd(modcmd_t cmd, void *data)
{
	int error = 0;
#ifdef _MODULE
	int bmajor = -1;
	int cmajor = -1;
#endif

	switch (cmd) {
	case MODULE_CMD_INIT:
#ifdef DEBUG
		logLevel = LOG_INFO;
#endif

		error = filemon_load(data);
#ifdef _MODULE
		if (!error)
			error = devsw_attach("filemon", NULL, &bmajor,
			    &filemon_cdevsw, &cmajor);
#endif
		break;

	case MODULE_CMD_FINI:
		error = filemon_unload();
#ifdef _MODULE
		if (!error)
			error = devsw_detach(NULL, &filemon_cdevsw);
#endif
		break;

	case MODULE_CMD_STAT:
		log(LOG_INFO, "filemon: open=%d", n_open);
		break;

	default:
		error = ENOTTY;
		break;

	}

	return (error);
}
